File: sdk/typescript/src/secretEnv.ts
/**
 * @file secretEnv.ts
 * @description Seal an owner's secret env (today: the inference API key) to an
 * agent's agentSeal public key, so it never appears in clear in the envelope
 * the owner's wallet signs (issue #166).
 *
 * The sandbox "create" envelope is signed with EIP-191 `personal_sign`, and a
 * wallet shows that message verbatim. With the key in `env.API_KEY`, the
 * prompt displayed it, and every hop that relays the envelope (an
 * integrator's proxy, the attestor's job store, the sandbox provider) held it
 * in clear. Sealed instead, the envelope carries
 *
 *   env.SEAL_SECRET_ENV = base64( ECIES(agentSeal_pub, {"v":1,"owner",env}) )
 *
 * The owner still signs exactly the bytes the container receives; those bytes
 * are ciphertext that only agentSeal_priv opens (the sealed container after
 * `/provision`, and the attestor's KMS derivation). The container checks that
 * `owner` is the agent's live on-chain owner before applying it.
 *
 * ECIES is the eciesjs-compatible scheme the attestor and sealed already use
 * for `sealedKeys`: secp256k1 ECDH, HKDF-SHA256 over
 * `ephemeralPub(65) ‖ sharedPoint(65)` (no salt, no info), AES-256-GCM with a
 * 16-byte nonce; wire format `ephemeralPub(65) ‖ nonce(16) ‖ tag(16) ‖ ct`.
 */

import { secp256k1 } from '@noble/curves/secp256k1';
import { hkdf } from '@noble/hashes/hkdf';
import { sha256 } from '@noble/hashes/sha2';
import { gcm } from '@noble/ciphers/aes';
import { getAddress, hexToBytes, keccak256, toHex, type Address, type Hex } from 'viem';

/** `GET /config` value that says this attestor's sealed images open a secret env. */
export const SECRET_ENV_SCHEME = 'agent-seal-ecies-v1';
/** The container env var that carries the sealed secret env. */
export const SECRET_ENV_VAR = 'SEAL_SECRET_ENV';

/**
 * How `start` / `reset` / `retry` / `deploy` deliver the inference key:
 *  - `'auto'` (default): sealed to the agent when the attestor advertises
 *    `secret_env_scheme` and the agent exists already; otherwise in clear
 *    (older attestors, and a one-shot `deploy` with `sandbox`, whose agent
 *    does not exist yet when the envelope is signed).
 *  - `'sealed'`: sealed or throw — the wallet prompt never shows the key.
 *  - `'plaintext'`: legacy `env.API_KEY` in clear, for a sealed image built
 *    before `SEAL_SECRET_ENV` existed.
 */
export type SecretEnvMode = 'auto' | 'sealed' | 'plaintext';

type WebCrypto = { getRandomValues<T extends Uint8Array>(x: T): T };

/** Cryptographic random bytes: WebCrypto in browsers and Node >= 19, the
 *  node:crypto fallback on Node 18 (the engines floor). */
export function randomBytes(n: number): Uint8Array {
  const a = new Uint8Array(n);
  let cryptoObj = (globalThis as { crypto?: WebCrypto }).crypto;
  if (!cryptoObj?.getRandomValues && typeof require === 'function') {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    cryptoObj = (require('crypto') as { webcrypto?: WebCrypto }).webcrypto;
  }
  if (!cryptoObj?.getRandomValues) {
    throw new Error('AgenticID SDK: no WebCrypto available (need a browser or Node >= 18 with crypto.webcrypto)');
  }
  cryptoObj.getRandomValues(a);
  return a;
}

function base64(bytes: Uint8Array): string {
  const g = globalThis as { btoa?: (d: string) => string };
  if (typeof g.btoa === 'function') {
    let bin = '';
    for (const b of bytes) bin += String.fromCharCode(b);
    return g.btoa(bin);
  }
  return Buffer.from(bytes).toString('base64');
}

/** Ethereum address of a secp256k1 public key (compressed or uncompressed). */
export function publicKeyToAddress(pubkey: Uint8Array): Address {
  const uncompressed = secp256k1.ProjectivePoint.fromHex(pubkey).toRawBytes(false);
  return getAddress(`0x${keccak256(uncompressed.slice(1)).slice(-40)}`);
}

/** eciesjs-compatible ECIES encryption to a secp256k1 public key (33 or 65 bytes). */
export function eciesEncrypt(recipientPubkey: Uint8Array, plaintext: Uint8Array): Uint8Array {
  // fromHex validates the point is on the curve (throws otherwise).
  const recipient = secp256k1.ProjectivePoint.fromHex(recipientPubkey);
  const ephemeralPriv = randomValidPriv();
  const ephemeralPub = secp256k1.getPublicKey(ephemeralPriv, false);
  const shared = secp256k1.getSharedSecret(ephemeralPriv, recipient.toRawBytes(true), false);
  const ikm = new Uint8Array(ephemeralPub.length + shared.length);
  ikm.set(ephemeralPub, 0);
  ikm.set(shared, ephemeralPub.length);
  const key = hkdf(sha256, ikm, undefined, undefined, 32);
  const nonce = randomBytes(16);
  const sealed = gcm(key, nonce).encrypt(plaintext); // ciphertext ‖ tag(16)
  const ct = sealed.subarray(0, sealed.length - 16);
  const tag = sealed.subarray(sealed.length - 16);
  const out = new Uint8Array(65 + 16 + 16 + ct.length);
  out.set(ephemeralPub, 0);
  out.set(nonce, 65);
  out.set(tag, 81);
  out.set(ct, 97);
  return out;
}

function randomValidPriv(): Uint8Array {
  // A uniformly random 32-byte string is a valid scalar with overwhelming
  // probability; loop for the (~2^-128) remainder rather than bias it.
  for (;;) {
    const k = randomBytes(32);
    if (secp256k1.utils.isValidPrivateKey(k)) return k;
  }
}

/**
 * Seal `env` for one agent. `agentSealPubkey` is the agent's agentSeal public
 * key (hex, compressed or uncompressed); `owner` is the address that will sign
 * the envelope (the container requires it to be the agent's on-chain owner).
 * Returns the base64 value for `env.SEAL_SECRET_ENV`.
 */
export function sealSecretEnv(agentSealPubkey: Hex, owner: Address, env: Record<string, string>): string {
  const plaintext = new TextEncoder().encode(
    JSON.stringify({ v: 1, owner: getAddress(owner), env }),
  );
  return base64(eciesEncrypt(hexToBytes(agentSealPubkey), plaintext));
}

/** Parse and validate an agentSeal public key served by the attestor; returns
 *  its canonical compressed hex. Throws on anything that is not a curve point. */
export function parseAgentSealPubkey(pubkey: string): Hex {
  if (!/^0x([0-9a-fA-F]{66}|[0-9a-fA-F]{130})$/.test(pubkey)) {
    throw new Error('agent_seal_pubkey must be a 33- or 65-byte hex secp256k1 public key');
  }
  return toHex(secp256k1.ProjectivePoint.fromHex(hexToBytes(pubkey as Hex)).toRawBytes(true));
}

File: sealed/internal/secretenv/secretenv.go
// Package secretenv opens the owner's secret env (today: the inference API
// key) that the SDK sealed to this agent's agentSeal public key (issue #166).
//
// Before this channel the key rode the owner-signed sandbox "create"
// envelope as `env.API_KEY` in clear: the owner's wallet prompt displayed it,
// and every hop (an integrator's proxy, the attestor's job store, the sandbox
// provider) held it in plain text. Now the envelope carries only
//
//	SEAL_SECRET_ENV = base64( ECIES(agentSeal_pub, plaintext) )
//
// so the owner still signs the exact bytes the container receives, but those
// bytes are ciphertext. Only holders of agentSeal_priv can open it: this
// container (after Phase 1 /provision) and the attestor's KMS derivation.
//
// ECIES is the scheme sealed already uses for sealedKeys (eciesjs-compatible:
// secp256k1, HKDF-SHA256, AES-256-GCM with a 16-byte nonce).
//
// Plaintext (JSON, v1):
//
//	{"v":1,"owner":"0x…","env":{"API_KEY":"…"}}
//
// `owner` is the envelope signer. Open requires it to equal the agent's live
// on-chain owner, so a ciphertext signed by a previous owner cannot be
// replayed into a container the next owner controls.
package secretenv

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strings"
	"time"

	eciesgo "github.com/ecies/go/v2"
	"github.com/ethereum/go-ethereum/common"
)

// EnvVar is the container env var that carries the sealed secret env.
const EnvVar = "SEAL_SECRET_ENV"

// Version is the only plaintext version this binary understands.
const Version = 1

// supported lists the names a secret env may set. Unknown names are ignored
// (reported, never applied), so an SDK that learns a new name does not break
// an older image: the names it does know still apply.
var supported = map[string]bool{
	"API_KEY": true,
}

// Reason classes for an Open failure. They name a cause, never a value, so
// they are safe in a status report that leaves the container.
const (
	ReasonMalformed          = "malformed"
	ReasonUnsupportedVersion = "unsupported_version"
	ReasonNotSealedToAgent   = "not_sealed_to_this_agent"
	ReasonOwnerUnknown       = "owner_unknown"
	ReasonOwnerMismatch      = "owner_mismatch"
	ReasonInternal           = "internal"
)

// Error is an Open failure with its reason class.
type Error struct {
	Reason string
	err    error
}

func (e *Error) Error() string { return e.err.Error() }
func (e *Error) Unwrap() error { return e.err }

func fail(reason string, err error) error { return &Error{Reason: reason, err: err} }

// Reason returns the reason class of an Open error (ReasonMalformed for an
// error Open did not classify).
func Reason(err error) string {
	var e *Error
	if errors.As(err, &e) {
		return e.Reason
	}
	return ReasonMalformed
}

// Opened is the validated content of a sealed secret env.
type Opened struct {
	// Env holds only supported names with non-empty values.
	Env map[string]string
	// Ignored lists unsupported names (sorted), for a log line. Never values.
	Ignored []string
}

type payload struct {
	V     int               `json:"v"`
	Owner string            `json:"owner"`
	Env   map[string]string `json:"env"`
}

// Open decodes, decrypts and validates a sealed secret env.
//
// chainOwner is the agent's live on-chain owner (hex). An empty chainOwner
// fails closed: without it the owner binding cannot be checked.
//
// Errors never include plaintext values.
func Open(encoded string, agentSealPriv []byte, chainOwner string) (*Opened, error) {
	encoded = strings.TrimSpace(encoded)
	if encoded == "" {
		return nil, fail(ReasonMalformed, errors.New("empty"))
	}
	ct, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return nil, fail(ReasonMalformed, fmt.Errorf("base64: %w", err))
	}
	if len(agentSealPriv) != 32 {
		return nil, fail(ReasonInternal, fmt.Errorf("agentSeal key must be 32 bytes, got %d", len(agentSealPriv)))
	}
	pt, err := eciesgo.Decrypt(eciesgo.NewPrivateKeyFromBytes(agentSealPriv), ct)
	if err != nil {
		// Not sealed to this agent's key (wrong agent, or corrupted).
		return nil, fail(ReasonNotSealedToAgent, fmt.Errorf("ecies decrypt: %w", err))
	}

	var p payload
	dec := json.NewDecoder(bytes.NewReader(pt))
	dec.DisallowUnknownFields()
	if err := dec.Decode(&p); err != nil {
		return nil, fail(ReasonMalformed, errors.New("plaintext is not a v1 secret env document"))
	}
	if p.V != Version {
		return nil, fail(ReasonUnsupportedVersion, fmt.Errorf("unsupported version %d (want %d)", p.V, Version))
	}
	if !common.IsHexAddress(p.Owner) {
		return nil, fail(ReasonMalformed, errors.New("owner is not an address"))
	}
	if !common.IsHexAddress(chainOwner) {
		return nil, fail(ReasonOwnerUnknown, errors.New("on-chain owner unknown; refusing to apply an owner-bound secret"))
	}
	if common.HexToAddress(p.Owner) != common.HexToAddress(chainOwner) {
		return nil, fail(ReasonOwnerMismatch, fmt.Errorf("sealed for owner %s, but the agent's owner is %s",
			common.HexToAddress(p.Owner).Hex(), common.HexToAddress(chainOwner).Hex()))
	}

	out := &Opened{Env: map[string]string{}}
	for name, value := range p.Env {
		if !supported[name] {
			out.Ignored = append(out.Ignored, name)
			continue
		}
		if value == "" || strings.ContainsRune(value, 0) {
			return nil, fail(ReasonMalformed, fmt.Errorf("%s: empty or invalid value", name))
		}
		out.Env[name] = value
	}
	sort.Strings(out.Ignored)
	if len(out.Env) == 0 {
		return nil, fail(ReasonMalformed, errors.New("no supported names"))
	}
	return out, nil
}

// NotAppliedPrefix starts the status detail ResolveAPIKey returns when an
// owner supplied a sealed key that did not apply.
const NotAppliedPrefix = "secret_env_not_applied: "

// ResolveAPIKey returns the inference key the adapters should use, given the
// plain API_KEY env value and the raw SEAL_SECRET_ENV value, plus an issue
// for the attestor (empty when there is nothing to report):
//
//   - no secret env: plainKey, unchanged;
//   - a secret env that opens (owner-bound) and carries API_KEY: that value,
//     which wins over plainKey;
//   - a secret env that fails to open: plainKey, with a FAIL line. If that
//     leaves no key at all, the agent boots but cannot call its model, so
//     issue is "secret_env_not_applied: <reason>" (a reason class, never a
//     value) for the caller to report as an owner-recoverable warning.
//
// logf lines name variables and lengths only, never values.
func ResolveAPIKey(plainKey, encoded string, agentSealPriv []byte, chainOwner string, logf func(string, ...any)) (key, issue string) {
	if encoded == "" {
		return plainKey, ""
	}
	opened, err := Open(encoded, agentSealPriv, chainOwner)
	if err != nil {
		logf("FAIL %s not applied: %v", EnvVar, err)
		if plainKey == "" {
			issue = NotAppliedPrefix + Reason(err)
		}
		return plainKey, issue
	}
	for _, name := range opened.Ignored {
		logf("warn: %s: ignoring unsupported name %q", EnvVar, name)
	}
	key = opened.Env["API_KEY"]
	if key == "" {
		return plainKey, ""
	}
	if plainKey != "" {
		logf("warn: both API_KEY and %s are set; using the sealed value", EnvVar)
	}
	logf("OK   API_KEY (from %s, owner-bound): <set, %d chars>", EnvVar, len(key))
	return key, ""
}

// LiveOwner returns known when it is set. Otherwise (the boot-time ownerOf
// read failed, e.g. on a transient RPC error) it re-reads the owner up to
// attempts times, waiting backoff*n before try n, and returns "" if every
// read fails. Open fails closed on an unknown owner, so without the re-read
// one RPC error at boot would leave the agent without its sealed key.
func LiveOwner(known string, read func() (string, error), attempts int, backoff time.Duration, logf func(string, ...any)) string {
	if known != "" || read == nil {
		return known
	}
	for i := 1; i <= attempts; i++ {
		time.Sleep(backoff * time.Duration(i))
		owner, err := read()
		if err == nil && owner != "" {
			return owner
		}
		if err == nil {
			err = errors.New("empty owner")
		}
		logf("warn: %s: owner re-read %d/%d failed: %v", EnvVar, i, attempts, err)
	}
	return ""
}
